File: semestre-2/disciplina-3-RENOMEAR/unidade-3/aula-5/aula-5.py
import pandas as pd
from datetime import datetime

# ==============================================================================
# 1. EXTRAÇÃO (E) - Simulando Dados de Múltiplas Fontes (ODS)
# ==============================================================================

# Fonte 1: Dados brutos de vendas do Site (com alguns erros)
dados_site_brutos = {
    'id_transacao': [101, 102, 103, 104, 105],
    'produto': ['Notebook X', 'MousePad', 'Fone Pro', 'Teclado Mec', 'MousePad'],
    'valor_bruto': [1500.50, 50.00, 'R$ 200,99', 450.00, 50.00], # Erro: formato de moeda
    'localizacao': ['SP', 'RJ', 'sp', 'MG', 'rj'], # Erro: inconsistência de caixa
    'data_compra': ['2024-10-15', '2024-10-16', '2024-10-16', '2024-10-17', '2024-10-17'],
    'status_entrega': ['ENTREGUE', 'PENDENTE', 'ENTREGUE', 'ENTREGUE', 'PENDENTE']
}
df_site = pd.DataFrame(dados_site_brutos)

# Fonte 2: Dados de Avaliações do Aplicativo
dados_app_avaliacao = {
    'id_produto': [1003, 1002, 1004],
    'nome_produto': ['Fone Pro', 'MousePad', 'Teclado Mec'],
    'avaliacao': [5, 4, 5]
}
df_avaliacao = pd.DataFrame(dados_app_avaliacao)

print("--- 1. Extração: Dados Brutos do Site ---")
print(df_site)

# ==============================================================================
# 2. METADADOS - Definindo a Estrutura (Governança de Dados)
# ==============================================================================

# Metadados de Negócio e Técnicos para o Data Warehouse
metadados_dw = {
    'Tabela_Fato_Vendas': {
        'coluna': 'valor_venda_liquido',
        'tipo': 'Float',
        'regra_transformacao': 'Valor bruto - 10% de imposto',
        'descricao_negocio': 'Valor final da transação após impostos.'
    },
    'Tabela_Dim_Local': {
        'coluna': 'uf',
        'tipo': 'String',
        'regra_transformacao': 'Converter para Maiúsculas e Padronizar. Ex: SP',
        'descricao_negocio': 'Unidade Federativa padronizada para análise geográfica.'
    }
}

print("\n--- 2. Metadados: Regra para Valor de Venda Líquido ---")
print(f"Coluna: {metadados_dw['Tabela_Fato_Vendas']['coluna']}")
print(f"Regra: {metadados_dw['Tabela_Fato_Vendas']['regra_transformacao']}")

# ==============================================================================
# 3. TRANSFORMAÇÃO (T) - Limpeza e Padronização de Dados
# ==============================================================================

# Aplicando a "Regra de Transformação" do Metadado para 'valor_bruto'
def limpar_e_converter_valor(valor):
    try:
        # Remove caracteres indesejados e converte para float
        if isinstance(valor, str):
            valor = valor.replace('R$', '').replace(',', '.').strip()
        return float(valor)
    except:
        # Se falhar, retorna 0 ou um valor de erro (limpeza/tratamento de exceções)
        return 0.0

df_site['valor_limpo'] = df_site['valor_bruto'].apply(limpar_e_converter_valor)

# 3.1. Criação da Métrica (Cálculo de Fato) - Baseado no Metadado
IMPOSTO_PERCENTUAL = 0.10
df_site['valor_venda_liquido'] = df_site['valor_limpo'] * (1 - IMPOSTO_PERCENTUAL)

# 3.2. Padronização da Dimensão (Localização) - Baseado no Metadado
df_site['uf'] = df_site['localizacao'].str.upper()

# 3.3. Junção de Fontes (Integração de Dados)
# Simulando a necessidade de mapear o nome do produto para um ID
# Em um DW real, faríamos joins com tabelas de Dimensão Produto
df_site['id_produto'] = df_site['produto'].astype('category').cat.codes + 1001

# Selecionando colunas que iriam para o DW (Esquema Estrela/Fato)
df_fato = df_site[['id_transacao', 'id_produto', 'uf', 'data_compra', 'valor_venda_liquido']].copy()
df_fato['data_carga'] = datetime.now().strftime("%Y-%m-%d %H:%M:%S") # Metadado Operacional (Data de Carga)

print("\n--- 3. Transformação: Dados Finais para o DW (Fato Vendas) ---")
print(df_fato)

# ==============================================================================
# 4. CARGA (L) e ANÁLISE (OLAP/BI) - Simulação de Consulta
# ==============================================================================

# Em um ambiente real, 'df_fato' seria carregado no banco de dados (DW - Camada de Apresentação)

# Simulação de Consulta Analítica (OLAP) - Agregação Rápida
# Pergunta: Qual o total de vendas líquidas por UF? (Slice e Aggregate)
analise_vendas_uf = df_fato.groupby('uf')['valor_venda_liquido'].sum().reset_index()

print("\n--- 4. Análise: Vendas Líquidas Agregadas por UF ---")
print(analise_vendas_uf)

# Adicionando um exemplo de Data Mart (Vendas x Avaliação)
df_dim_prod = df_site[['id_produto', 'produto']].drop_duplicates()
df_analise_final = pd.merge(analise_vendas_uf, df_dim_prod, left_on='id_produto', right_on='id_produto', how='left')
# ... (O merge direto aqui seria complexo sem IDs consistentes, focaremos no exemplo anterior)

# Conclusão
print("\n✅ Processo ETL e Metadados simulado com sucesso!")
print(f"Total de Registros carregados (Fato Vendas): {len(df_fato)}")

File: semestre-2/disciplina-3-RENOMEAR/unidade-3/aula-5/aula-5.js
/**
 * Arquivo: aula-5.js
 * Tema: Arquitetura de Dados Corporativos, ETL, Modelagem Dimensional e Análise OLAP
 *
 * Simulação do Processo de Transformação e Análise de Dados de Vendas
 * para o Data Warehouse, focando em JavaScript.
 */

// ==============================================================================
// 1. CARGA (L) - Dados Simulados no DW (Modelagem Dimensional Simples)
//    (Estes dados representam a Tabela Fato 'Vendas' após o processo ETL)
// ==============================================================================
const dadosDW = [
    { id_transacao: 101, id_produto: 1001, uf: 'SP', valor_liquido: 1350.45, data: '2024-10-15' },
    { id_transacao: 102, id_produto: 1002, uf: 'RJ', valor_liquido: 45.00, data: '2024-10-16' },
    { id_transacao: 103, id_produto: 1003, uf: 'SP', valor_liquido: 180.89, data: '2024-10-16' },
    { id_transacao: 104, id_produto: 1004, uf: 'MG', valor_liquido: 405.00, data: '2024-10-17' },
    { id_transacao: 105, id_produto: 1002, uf: 'RJ', valor_liquido: 45.00, data: '2024-10-17' },
    { id_transacao: 106, id_produto: 1001, uf: 'SP', valor_liquido: 1350.45, data: '2024-10-18' },
];

// Tabela Dimensão Produto (Dimensão)
const dimProduto = {
    1001: { nome: 'Notebook X', categoria: 'Eletrônicos' },
    1002: { nome: 'MousePad', categoria: 'Acessórios' },
    1003: { nome: 'Fone Pro', categoria: 'Áudio' },
    1004: { nome: 'Teclado Mec', categoria: 'Periféricos' },
};

console.log("--- 1. Dados Carregados (Data Warehouse) ---");
console.log(`Total de Transações: ${dadosDW.length}`);

// ==============================================================================
// 2. SIMULAÇÃO DE OLAP (Online Analytical Processing)
//    Realizando consultas agregadas (Slice/Dice) usando JS
// ==============================================================================

/**
 * Função que simula a agregação de dados em um "Cubo OLAP".
 * Agrega o valor líquido das vendas por uma dimensão específica (ex: UF, Categoria).
 * @param {Array} data - A tabela Fato (dadosDW).
 * @param {string} dimensionKey - A chave da dimensão para agrupar (ex: 'uf').
 * @returns {Object} Um objeto com os totais agrupados.
 */
function aggregateByDimension(data, dimensionKey) {
    const aggregation = {};

    data.forEach(fato => {
        let key = fato[dimensionKey];

        // Se a dimensão for o produto (id_produto), usamos a dimensão externa para buscar a Categoria
        if (dimensionKey === 'categoria') {
            const produtoInfo = dimProduto[fato.id_produto];
            key = produtoInfo ? produtoInfo.categoria : 'Desconhecido';
        }

        // Soma o valor líquido (o "Fato" numérico)
        if (!aggregation[key]) {
            aggregation[key] = 0;
        }
        aggregation[key] += fato.valor_liquido;
    });

    // Formata os resultados
    return Object.keys(aggregation).map(key => ({
        [dimensionKey]: key,
        total_vendas_liquidas: aggregation[key].toFixed(2) // 2 casas decimais
    }));
}


// AÇÃO 1: Análise por Localização (Dimensão UF)
const analisePorUF = aggregateByDimension(dadosDW, 'uf');
console.log("\n--- 2.1. Análise OLAP: Vendas por Localização (UF) ---");
console.table(analisePorUF);

// AÇÃO 2: Análise por Categoria de Produto (Drill-down na Dimensão Produto)
const analisePorCategoria = aggregateByDimension(dadosDW, 'categoria');
console.log("\n--- 2.2. Análise OLAP: Vendas por Categoria de Produto ---");
console.table(analisePorCategoria);


// ==============================================================================
// 3. METADADOS - Uso e Consulta
// ==============================================================================

// Simulação de Metadados de Negócio (Descrição do Fato principal)
const metadadoVendas = {
    nome: "valor_liquido",
    tipo: "Métrica (Fato)",
    origem: "Calculado em ETL (Valor Bruto - Impostos)",
    definicao_negocio: "Receita final da empresa por transação, fundamental para o cálculo de Margem de Lucro."
};

console.log("\n--- 3. Consulta de Metadado de Negócio ---");
console.log(`Métrica: ${metadadoVendas.nome}`);
console.log(`Definição: ${metadadoVendas.definicao_negocio}`);


console.log("\n✅ Simulação de Carga e Análise OLAP em JavaScript concluída.");